src/signal_processing_algorithms/energy_statistics.py

```diff
@@ -0,0 +1,309 @@
+            - np.mean(distances_within_y)
+        )
+
+    @staticmethod
+    def get_t_from_e(t_coefficient: float, e_statistic: float) -> float:
+        """
+        Return T-statistic from E-statistic and T-coefficient.
+
+        :param t_coefficient: T-coefficient.
+        :param e_statistic: E-statistic.
+        :return:T-statistic.
+        """
+        return t_coefficient * e_statistic
+
+    @staticmethod
+    def get_h_from_e_and_distance_matrix(e: float, distances_between_xy: np.ndarray) -> float:
+        """
+        Return H (E-coefficient of inhomogeneity) from E-statistic and distance matrix.
+
+        :param e: E-statistic.
+        :param distances_between_xy: Distance matrix of pairwise distances between
+        distribution x and y.
+        :return:
+        """
+        avg_distances_between_xy = np.mean(distances_between_xy)
+        return float(e / (2 * avg_distances_between_xy)) if avg_distances_between_xy > 0 else 0.0
+
+    def _get_statistics_from_distance_matrices(
+        self,
+        distances_within_x: np.ndarray,
+        distances_within_y: np.ndarray,
+        distances_between_xy: np.ndarray,
+    ) -> Tuple[float, float, float]:
+        """
+        Return energy statistics.
+
+        :param distances_within_x: Distance matrix of pairwise distances within distribution x.
+        :param distances_within_y: Distance matrix of pairwise distances within distribution y.
+        :param distances_between_xy: Distance matrix of pairwise distances
+        between distribution x and y.
+        :return: E-statistic, T-statistic and H(E-coefficient of inhomogeneity)
+        """
+        e = self.get_e_from_distance_matrices(
+            distances_within_x, distances_within_y, distances_between_xy
+        )
+        t = self.get_t_from_e(self._t_coefficient, e)
+        h = self.get_h_from_e_and_distance_matrix(e, distances_between_xy)
+        return e, t, h
+
+    def _get_statistics_from_combined_distance_matrix(
+        self, distance_matrix: np.ndarray
+    ) -> Tuple[float, float, float]:
+        """
+        Return energy statistics for distance matrix.
+
+        :param distance_matrix: Distance matrix of pairwise distances within distribution formed by combining distributions x and y.
+        :return: E-statistic, T-statistic and H(E-coefficient of inhomogeneity)
+        """
+        len_distance_matrix = len(distance_matrix)
+        distances_within_x = distance_matrix[0 : self._n, 0 : self._n]
+        distances_within_y = distance_matrix[
+            self._n : len_distance_matrix, self._n : len_distance_matrix
+        ]
+        distances_between_xy = distance_matrix[self._n : len_distance_matrix, 0 : self._n]
+        return self._get_statistics_from_distance_matrices(
+            distances_within_x, distances_within_y, distances_between_xy
+        )
+
+    def get_statistics_and_probabilities(self, num_tests: int = 1000) -> EnergyStatisticsTestResult:
```

Comment:
This is looking good, but I think it will clean up the design/APIs a bit to separate concerns:
* difference matrix calculation (this is already separate)
* energy statistics calculation (based on x,y OR diff_matrix, n, m)
* permutation testing (is just shuffling stuff and otherwise using the above two for calculations)

Reply:
Done.